snapcraft/models/project.py

```diff
@@ -2218,6 +2209,205 @@ def get_partitions(self) -> list[str] | None:
+        default=None,
+        description="Not available for core22. Use the `architectures` key.",
+        exclude=True,
+        repr=False,
+    )
+
+    @pydantic.model_validator(mode="after")
+    def _validate_platforms_and_architectures(self) -> Self:
+        """Validate usage of platforms and architectures.
+
+        - can optionally define architectures
+        - cannot define platforms
+        """
+        # this is a one-shot - the value should not change when re-validating
+        if self.architectures and self._architectures_in_yaml is None:
+            # record if architectures are defined in the yaml for remote-build (#4881)
+            self._architectures_in_yaml = True
+
+        elif not self.architectures:
+            self._architectures_in_yaml = False
+            # set default value
+            host_arch = str(DebianArchitecture.from_host())
+            self.architectures = [
+                Architecture(
+                    build_on=[host_arch],
+                    build_for=[host_arch],
+                )
+            ]
+
+        return self
+
+    @pydantic.field_validator("architectures")
+    @classmethod
+    def _validate_architecture_data(
+        cls, architectures: list[str | Architecture], info: pydantic.ValidationInfo
+    ) -> list[Architecture]:
+        """Validate architecture data."""
+        return validate_architectures(architectures)
+
+
+class BareCore22Project(Core22Project):
+    base: Literal["bare"]  # type: ignore[assignment,reportIncompatibleVariableOverride]
+    build_base: Literal["core22"]  # type: ignore[reportIncompatibleVariableOverride]
+
+
+class BaseCore22Project(Core22Project):
+    type: Literal["base"]  # type: ignore[assignment,reportIncompatibleVariableOverride]
+    base: SkipJsonSchema[None] = None  # type: ignore[assignment,reportIncompatibleVariableOverride]
+    build_base: Literal["core22"]  # type: ignore[reportIncompatibleVariableOverride]
+
+
+class Core24Project(Project):
+    type: Annotated[  # type: ignore[assignment,reportIncompatibleVariableOverride]
+        Literal["app", "gadget", "kernel", "snapd", None],
+        _custom_error("Input should be 'app', 'base', 'gadget', 'kernel' or 'snapd'"),
+    ] = pydantic.Field(
+        default=None, description="The snap's type.", examples=["kernel"]
+    )
+
+    base: Literal["core24"]  # type: ignore[reportIncompatibleVariableOverride]
+
+    architectures: SkipJsonSchema[  # type: ignore[reportIncompatibleVariableOverride]
+        Annotated[
+            None,
+            _custom_error(
+                "'architectures' key is not supported for base 'core24'. Use 'platforms' key instead."
+            ),
+        ]
+    ] = pydantic.Field(
+        default=None,
+        description="The architectures key is only used in core22 snaps. For core24 and newer snaps, use the ``platforms`` key.",
```

Comment:
```suggestion
        description="The architectures key is only used in core22 snaps and below. For core24 and newer snaps, use the ``platforms`` key.",
```

I know this PR is ignoring core20 and lower, but this is might've caused some confusion in the future.